### ISSUE TITLE
ShellItem Properties - System.Runtime.InteropServices.COMException

### ISSUE THREAD

**Issue by gave92** [dahall/Vanara#126]
**Describe the bug**
Sorry, I'm bugging you again. The following code results in a System.Runtime.InteropServices.COMException.

**What code is involved**
```c#
var si = new ShellItem(@"C:\ANY_FILE_PATH");
var key = si.PropertyDescriptions.First().PropertyKey; ---> ok
var prop1 = si.Properties[key]; ---> exception
var prop2 = si.Properties.Keys; ----> exception
var prop3 = si.Properties.First(); ----> exception
```

**Stack trace**
> in Vanara.PInvoke.Shell32.IShellItem2.GetPropertyStore(GETPROPERTYSTOREFLAGS flags, Guid& riid)
   in Vanara.Windows.Shell.ShellItemPropertyStore.GetIPropertyStore()
   in Vanara.Windows.Shell.PropertyStore.Run[T](Func`2 action)
   in Vanara.Windows.Shell.PropertyStore.get_Keys()
   in FilesFullTrust.Program.Main(String[] args) in C:\Users\Marco\source\repos\files-uwp\Files.Launcher\Program.cs:riga 49

**Research**
The following code works instead:
```cs
var si = new ShellItem(@"C:\ANY_FILE_PATH");
string fileSize = GetFileStringProperty(si, Vanara.PInvoke.Ole32.PROPERTYKEY.System.ItemName);
....

private static string GetFileStringProperty(ShellItem folderItem, Vanara.PInvoke.Ole32.PROPERTYKEY propertyKey)
{
    var ps = ((Vanara.PInvoke.Shell32.IShellItem2)folderItem.IShellItem).GetPropertyStore(Vanara.PInvoke.PropSys.GETPROPERTYSTOREFLAGS.GPS_DEFAULT, typeof(Vanara.PInvoke.PropSys.IPropertyStore).GUID);
    var pv = new Vanara.PInvoke.Ole32.PROPVARIANT();
    ps.GetValue(propertyKey, pv);
     var pdesc = PropertyDescription.Create(propertyKey);
     var pvalue = pdesc?.FormatForDisplay(pv, Vanara.PInvoke.PropSys.PROPDESC_FORMAT_FLAGS.PDFF_DEFAULT);
     pdesc?.Dispose();
     pv.Dispose();
     Marshal.ReleaseComObject(ps);
     return pvalue;
}
```

**Comment by dahall** [comment]
You can enumerate all valid properties using the `Properties.Keys` property. When I queried against a Word document, "System.Size" wasn't one of the properties.

**Comment by gave92** [comment]
I updated the first post. Should have described the issue better, thanks for looking into this

**Comment by dahall** [comment]
I also just tested and "System.Size" works as a R/O store property. After some reading, it appears that the R/O and R/W stores are actually two different streams. If you allow for inheritance, then there are multiple streams. Any thoughts on how to make this simpler to use?

**Comment by gave92** [comment]
I've tried your suggestion and works perfectly, thanks.

As for your question.. I'm not sure. I admit I just found out about the existence of R/W and R/O stores.
Perhaps you could make `ReadOnly = true` the default? If you want to write a property to the store you will have to explicitly set `ReadOnly = false`, then the library will try to open the R/W store and fail with a descriptive error if no R/W stores exist.

**Comment by dahall** [comment]
Yes, items can have an unlimited number of stores based on inheritance of containers. Most files just have one or two (R/O, R/W). I was trying to hide that complexity in my implementation, but obviously failed.

**Comment by dahall** [comment]
I'll post here if I decide to change the implementation model.

**Comment by gave92** [comment]
A more descriptive reason for failure would surely be of help in any case :)

Btw what is the correct way of disposing of the property store? And in general am I supposed to dispose of ShellFolder items returned when enumerating the contents of a ShellFolder?
```cs
var recycler = new ShellFolder(Vanara.PInvoke.Shell32.KNOWNFOLDERID.FOLDERID_RecycleBinFolder);
foreach (var folderItem in recycler)
{
	try
	{
		folderItem.Properties.ReadOnly = true;
		folderItem.Properties.NoInheritedProperties = false;
		string fileSize = folderItem.Properties.GetPropertyString(Vanara.PInvoke.Ole32.PROPERTYKEY.System.Size);		
	}
	catch (System.IO.FileNotFoundException)
	{
		// Happens if files are being deleted
	}
	finally
	{
		folderItem?.Properties?.Dispose(); -----> should I do this?
		folderItem?.Dispose(); -----> should I do this?
	}
}
```
If I don't dispose and try to delete the item from Explorer the file delete dialog will hang until I close the program. Sorry for the dumb question.

**Comment by dahall** [comment]
Yes for `folderItem` and no for `folderItem.Properties` since this is done automatically by `folderItem`'s disposal.

**Comment by dahall** [comment]
So I did a few things to fix/compensate on your original issue submissions:
1. Added improved exception (`InvalidOperationException`) when IPropertyStore instance cannot be created due to a shell item with no r/w support.
2. Changed default behavior of ShellItem.Properties so that it is read-only. It was failing for all shell items that didn't have their own properties. To get previous behavior, simply set "shellItem.Properties.ReadOnly = false" before setting properties.
3. Added notes to `ShellItem.Properties` to explain how to use.
